File: src/nickname_resolver.py
"""Nickname resolution for matching informal names to formal names."""

import json
import os
from pathlib import Path
from typing import Dict, Set, Optional

from rapidfuzz.distance import JaroWinkler


class NicknameResolver:
    """Resolves nicknames to formal names and vice versa."""
    
    # Default threshold for fuzzy matching (0.85 = 85% similarity)
    DEFAULT_FUZZY_THRESHOLD = 0.85
    
    def __init__(self, nicknames_file: Optional[str] = None):
        """Initialize the resolver with nickname mappings.
        
        Args:
            nicknames_file: Path to the nicknames JSON file. If None, uses default location.
        """
        if nicknames_file is None:
            # Default location: config/nicknames.json relative to project root
            project_root = Path(__file__).parent.parent
            nicknames_file = project_root / "config" / "nicknames.json"
        
        self._nickname_to_formal: Dict[str, Set[str]] = {}
        self._formal_to_nicknames: Dict[str, Set[str]] = {}
        self._all_names: Set[str] = set()
        
        self._load_nicknames(nicknames_file)
    
    def _load_nicknames(self, filepath: str | Path) -> None:
        """Load nickname mappings from JSON file."""
        filepath = Path(filepath)
        
        if not filepath.exists():
            # No nicknames file - resolver will still work with fuzzy matching
            return
        
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                data = json.load(f)
        except (json.JSONDecodeError, IOError):
            # Invalid file - continue without nicknames
            return
        
        # Build bidirectional lookup
        for nickname, formal_names in data.items():
            nickname_lower = nickname.lower()
            self._all_names.add(nickname_lower)
            
            if nickname_lower not in self._nickname_to_formal:
                self._nickname_to_formal[nickname_lower] = set()
            
            for formal in formal_names:
                formal_lower = formal.lower()
                self._all_names.add(formal_lower)
                
                # nickname -> formal mapping
                self._nickname_to_formal[nickname_lower].add(formal_lower)
                
                # formal -> nickname mapping
                if formal_lower not in self._formal_to_nicknames:
                    self._formal_to_nicknames[formal_lower] = set()
                self._formal_to_nicknames[formal_lower].add(nickname_lower)
                
                # Also treat formal names as potentially mapping to each other
                # e.g., "nicholas" maps to "nick" and "nico", so "nicholas" group includes those
    
    def get_formal_names(self, nickname: str) -> Set[str]:
        """Get all formal names that could match a nickname.
        
        Args:
            nickname: The nickname to look up.
            
        Returns:
            Set of formal names that this nickname maps to.
        """
        return self._nickname_to_formal.get(nickname.lower(), set())
    
    def get_nicknames(self, formal_name: str) -> Set[str]:
        """Get all nicknames for a formal name.
        
        Args:
            formal_name: The formal name to look up.
            
        Returns:
            Set of nicknames for this formal name.
        """
        return self._formal_to_nicknames.get(formal_name.lower(), set())
    
    def get_all_equivalents(self, name: str) -> Set[str]:
        """Get all equivalent names (both directions).
        
        Args:
            name: The name to find equivalents for.
            
        Returns:
            Set of all names that could be equivalent to this one.
        """
        name_lower = name.lower()
        equivalents = {name_lower}
        
        # Add formal names if this is a nickname
        formal_names = self.get_formal_names(name_lower)
        equivalents.update(formal_names)
        
        # Add nicknames if this is a formal name
        nicknames = self.get_nicknames(name_lower)
        equivalents.update(nicknames)
        
        # For each formal name, also add other nicknames
        for formal in formal_names:
            equivalents.update(self.get_nicknames(formal))
        
        # For each nickname, also add other formal names
        for nick in nicknames:
            equivalents.update(self.get_formal_names(nick))
        
        return equivalents
    
    def are_names_equivalent(self, name1: str, name2: str) -> bool:
        """Check if two names are equivalent via nickname mapping.
        
        Args:
            name1: First name to compare.
            name2: Second name to compare.
            
        Returns:
            True if the names are equivalent.
        """
        name1_lower = name1.lower()
        name2_lower = name2.lower()
        
        # Exact match
        if name1_lower == name2_lower:
            return True
        
        # Check if name2 is in name1's equivalents
        return name2_lower in self.get_all_equivalents(name1_lower)
    
    def fuzzy_match(self, name1: str, name2: str, threshold: float = None) -> bool:
        """Check if two names are similar using fuzzy matching.
        
        Uses Jaro-Winkler similarity which is good for names.
        
        Args:
            name1: First name to compare.
            name2: Second name to compare.
            threshold: Minimum similarity score (0-1). Defaults to DEFAULT_FUZZY_THRESHOLD.
            
        Returns:
            True if the names are similar enough.
        """
        if threshold is None:
            threshold = self.DEFAULT_FUZZY_THRESHOLD
        
        # Jaro-Winkler is good for names - gives higher scores to strings
        # that match from the beginning
        similarity = JaroWinkler.normalized_similarity(name1.lower(), name2.lower())
        return similarity >= threshold
    
    def get_fuzzy_score(self, name1: str, name2: str) -> float:
        """Get the fuzzy similarity score between two names.
        
        Args:
            name1: First name to compare.
            name2: Second name to compare.
            
        Returns:
            Similarity score between 0 and 1.
        """
        return JaroWinkler.normalized_similarity(name1.lower(), name2.lower())


# Global instance for convenience
_resolver: Optional[NicknameResolver] = None


def get_resolver() -> NicknameResolver:
    """Get the global NicknameResolver instance."""
    global _resolver
    if _resolver is None:
        _resolver = NicknameResolver()
    return _resolver


def are_names_equivalent(name1: str, name2: str) -> bool:
    """Check if two names are equivalent using nickname mappings.
    
    Convenience function using the global resolver.
    """
    return get_resolver().are_names_equivalent(name1, name2)


def fuzzy_match(name1: str, name2: str, threshold: float = None) -> bool:
    """Check if two names are similar using fuzzy matching.
    
    Convenience function using the global resolver.
    """
    return get_resolver().fuzzy_match(name1, name2, threshold)


def get_fuzzy_score(name1: str, name2: str) -> float:
    """Get the fuzzy similarity score between two names.
    
    Convenience function using the global resolver.
    """
    return get_resolver().get_fuzzy_score(name1, name2)

File: packages/core/src/nickname-resolver.ts
/**
 * Nickname Resolver Module
 * Handles bidirectional nickname-to-formal name lookups and fuzzy name matching
 *
 * This module provides:
 * - Bidirectional lookup: nickname -> formal names and formal name -> nicknames
 * - Transitive equivalence: "Nick" -> "Nicholas" -> "Nico" means Nick ≡ Nico
 * - Jaro-Winkler fuzzy matching for handling typos and variations
 */

import nicknames from './data/nicknames.json';

// Default fuzzy match threshold (0.85 = 85% similarity required)
export const DEFAULT_FUZZY_THRESHOLD = 0.85;

/**
 * NicknameResolver provides bidirectional nickname-to-formal name lookups
 * and fuzzy name matching using Jaro-Winkler similarity.
 *
 * The nickname data structure in nicknames.json maps nicknames to formal names:
 * { "nick": ["nicholas", "nicolas", "nico"], ... }
 *
 * This resolver builds bidirectional mappings for efficient lookup in both directions.
 */
export class NicknameResolver {
  private nicknameToFormal: Map<string, Set<string>>;
  private formalToNickname: Map<string, Set<string>>;
  private allNames: Set<string>;

  constructor() {
    this.nicknameToFormal = new Map();
    this.formalToNickname = new Map();
    this.allNames = new Set();
    this.loadNicknames();
  }

  /**
   * Load nicknames from the JSON data and build bidirectional mappings
   */
  private loadNicknames(): void {
    // nicknames.json format: { "nickname": ["formal1", "formal2", ...] }
    for (const [nickname, formals] of Object.entries(nicknames)) {
      const nicknameLower = nickname.toLowerCase();
      this.allNames.add(nicknameLower);

      // Initialize nickname -> formal mapping if not exists
      if (!this.nicknameToFormal.has(nicknameLower)) {
        this.nicknameToFormal.set(nicknameLower, new Set());
      }

      for (const formal of formals as string[]) {
        const formalLower = formal.toLowerCase();
        this.allNames.add(formalLower);

        // Add nickname -> formal mapping
        this.nicknameToFormal.get(nicknameLower)!.add(formalLower);

        // Add formal -> nickname mapping
        if (!this.formalToNickname.has(formalLower)) {
          this.formalToNickname.set(formalLower, new Set());
        }
        this.formalToNickname.get(formalLower)!.add(nicknameLower);
      }
    }
  }

  /**
   * Get formal names for a nickname
   * @param nickname The nickname to look up (e.g., "Nick")
   * @returns Set of formal names (e.g., {"nicholas", "nicolas", "nico"})
   */
  getFormalNames(nickname: string): Set<string> {
    return this.nicknameToFormal.get(nickname.toLowerCase()) ?? new Set();
  }

  /**
   * Get nicknames for a formal name
   * @param formalName The formal name to look up (e.g., "Michael")
   * @returns Set of nicknames (e.g., {"mike"})
   */
  getNicknames(formalName: string): Set<string> {
    return this.formalToNickname.get(formalName.toLowerCase()) ?? new Set();
  }

  /**
   * Get all equivalent names (formal + nicknames + name itself)
   * This provides transitive closure - if Nick -> Nicholas and Nicholas -> Nico,
   * then Nick, Nicholas, and Nico are all equivalents of each other.
   *
   * @param name Any name (nickname or formal)
   * @returns Set of all equivalent names
   */
  getAllEquivalents(name: string): Set<string> {
    const nameLower = name.toLowerCase();
    const equivalents = new Set<string>([nameLower]);

    // Add formal names if this is a nickname
    const formals = this.getFormalNames(nameLower);
    formals.forEach((f) => equivalents.add(f));

    // Add nicknames if this is a formal name
    const nicks = this.getNicknames(nameLower);
    nicks.forEach((n) => equivalents.add(n));

    // For each formal name found, also add its other nicknames (transitive closure)
    formals.forEach((formal) => {
      const otherNicks = this.getNicknames(formal);
      otherNicks.forEach((n) => equivalents.add(n));
    });

    // For each nickname found, also add other formal names (transitive closure)
    nicks.forEach((nick) => {
      const otherFormals = this.getFormalNames(nick);
      otherFormals.forEach((f) => equivalents.add(f));
    });

    return equivalents;
  }

  /**
   * Check if two names are equivalent (same person)
   * @param name1 First name
   * @param name2 Second name
   * @returns true if names are equivalent
   */
  areNamesEquivalent(name1: string, name2: string): boolean {
    const name1Lower = name1.toLowerCase();
    const name2Lower = name2.toLowerCase();

    if (name1Lower === name2Lower) return true;

    const equivalents = this.getAllEquivalents(name1);
    return equivalents.has(name2Lower);
  }

  /**
   * Calculate Jaro similarity score between two strings
   * @param s1 First string
   * @param s2 Second string
   * @returns Jaro similarity score between 0 and 1
   */
  private jaroSimilarity(s1: string, s2: string): number {
    const str1 = s1.toLowerCase();
    const str2 = s2.toLowerCase();

    if (str1 === str2) return 1.0;
    if (str1.length === 0 || str2.length === 0) return 0.0;

    // Calculate match window
    const matchDistance = Math.floor(Math.max(str1.length, str2.length) / 2) - 1;
    const s1Matches = new Array(str1.length).fill(false);
    const s2Matches = new Array(str2.length).fill(false);

    let matches = 0;
    let transpositions = 0;

    // Find matching characters
    for (let i = 0; i < str1.length; i++) {
      const start = Math.max(0, i - matchDistance);
      const end = Math.min(i + matchDistance + 1, str2.length);

      for (let j = start; j < end; j++) {
        if (s2Matches[j] || str1[i] !== str2[j]) continue;
        s1Matches[i] = true;
        s2Matches[j] = true;
        matches++;
        break;
      }
    }

    if (matches === 0) return 0.0;

    // Count transpositions
    let k = 0;
    for (let i = 0; i < str1.length; i++) {
      if (!s1Matches[i]) continue;
      while (!s2Matches[k]) k++;
      if (str1[i] !== str2[k]) transpositions++;
      k++;
    }

    // Calculate Jaro similarity
    const jaro =
      (matches / str1.length + matches / str2.length + (matches - transpositions / 2) / matches) /
      3;

    return jaro;
  }

  /**
   * Calculate Jaro-Winkler similarity score between two strings
   * This algorithm is specifically chosen for name matching because:
   * - It gives higher scores to strings that match from the beginning
   * - It handles common typos and character transpositions well
   * - It is effective for both short names and longer full names
   *
   * @param s1 First string
   * @param s2 Second string
   * @returns Similarity score between 0 and 1
   */
  getFuzzyScore(s1: string, s2: string): number {
    const str1 = s1.toLowerCase();
    const str2 = s2.toLowerCase();

    if (str1 === str2) return 1.0;
    if (str1.length === 0 || str2.length === 0) return 0.0;

    // Get Jaro similarity
    const jaro = this.jaroSimilarity(str1, str2);

    // Winkler modification - boost for common prefix (up to 4 characters)
    let prefix = 0;
    const maxPrefix = Math.min(4, str1.length, str2.length);
    for (let i = 0; i < maxPrefix; i++) {
      if (str1[i] === str2[i]) prefix++;
      else break;
    }

    // Standard Winkler scaling factor is 0.1
    const scalingFactor = 0.1;
    return jaro + prefix * scalingFactor * (1 - jaro);
  }

  /**
   * Check if two names match using fuzzy matching
   * @param name1 First name
   * @param name2 Second name
   * @param threshold Minimum similarity score (default 0.85)
   * @returns true if names are similar enough
   */
  fuzzyMatch(name1: string, name2: string, threshold = DEFAULT_FUZZY_THRESHOLD): boolean {
    return this.getFuzzyScore(name1, name2) >= threshold;
  }

  /**
   * Check if a name is known (either as a nickname or formal name)
   * @param name The name to check
   * @returns true if the name exists in the nickname database
   */
  isKnownName(name: string): boolean {
    return this.allNames.has(name.toLowerCase());
  }

  /**
   * Get all known names in the database
   * @returns Set of all names (both nicknames and formal names)
   */
  getAllKnownNames(): Set<string> {
    return new Set(this.allNames);
  }
}

// Export a default instance for convenience
export const nicknameResolver = new NicknameResolver();

// Export convenience functions that use the default resolver instance
export function getFormalNames(nickname: string): Set<string> {
  return nicknameResolver.getFormalNames(nickname);
}

export function getNicknames(formalName: string): Set<string> {
  return nicknameResolver.getNicknames(formalName);
}

export function getAllEquivalents(name: string): Set<string> {
  return nicknameResolver.getAllEquivalents(name);
}

export function areNamesEquivalent(name1: string, name2: string): boolean {
  return nicknameResolver.areNamesEquivalent(name1, name2);
}

export function getFuzzyScore(name1: string, name2: string): number {
  return nicknameResolver.getFuzzyScore(name1, name2);
}

export function fuzzyMatch(
  name1: string,
  name2: string,
  threshold = DEFAULT_FUZZY_THRESHOLD
): boolean {
  return nicknameResolver.fuzzyMatch(name1, name2, threshold);
}
